datafusion/core/tests/physical_optimizer/sanity_checker.rs

```diff
@@ -379,6 +379,32 @@ async fn test_analyzer() -> Result<()> {
     Ok(())
 }
 
+#[tokio::test]
+async fn test_order_by_case_requirement() -> Result<()> {
+    let sql = r#"
+    WITH typ(oid, typnamespace, typname, typtype) AS (
+                      SELECT * FROM (VALUES (1, 10, 't1', 'b'))
+            UNION ALL SELECT * FROM (VALUES (2, NULL, 't2', 'b'))
+            UNION ALL SELECT * FROM (VALUES (3, 12, 't3', NULL))
```

Comment:
Nit: maybe adding a non-NULL value which is retained by the filter but not matching any `WHEN` branch like `m`? We are exercising the implicit `ELSE null` with rows having `typtype` as `NULL` value, but it might be worth checking with explicit values too?